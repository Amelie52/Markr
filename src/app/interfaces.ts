export type Tag = {
  label: string;
  value: string;
};

export interface Bookmark {
  id: string;
  url: string;
  title: string;
  author?: string;
  createdDate: Date;
  tags: Tag[];
  width: number;
  height: number;
  duration?: number;
}
