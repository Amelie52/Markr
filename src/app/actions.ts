import { Bookmark, Tag } from './interfaces';

export enum BOOKMARK {
  CREATE = 'app/BOOKMARK/CREATE',
  EDIT = 'app/BOOKMARK/EDIT',
  DELETE = 'app/BOOKMARK/DELETE',
}

export enum TAGS {
  CREATE = 'app/TAGS/CREATE',
}

export interface CreateBookmarkAction {
  type: BOOKMARK.CREATE;
  bookmark: Bookmark;
  tags: Tag[];
}
export interface EditBookmarkAction {
  type: BOOKMARK.EDIT;
  bookmarkId: string;
  tags: Tag[];
}
export interface DeleteBookmarkAction {
  type: BOOKMARK.DELETE;
  bookmarkId: string;
}

export interface CreateTagsAction {
  type: TAGS.CREATE;
  tags: Tag[];
}

export type Action =
  | CreateBookmarkAction
  | EditBookmarkAction
  | DeleteBookmarkAction
  | CreateTagsAction;

export function createBookmark(bookmark: Bookmark, tags: Tag[]): CreateBookmarkAction {
  return {
    type: BOOKMARK.CREATE,
    bookmark,
    tags,
  };
}

export function editBookmark(bookmarkId: string, tags: Tag[]): EditBookmarkAction {
  return {
    type: BOOKMARK.EDIT,
    bookmarkId,
    tags,
  };
}

export function deleteBookmark(bookmarkId: string): DeleteBookmarkAction {
  return {
    type: BOOKMARK.DELETE,
    bookmarkId,
  };
}

export function createTags(tags: Tag[]): CreateTagsAction {
  return {
    type: TAGS.CREATE,
    tags,
  };
}
