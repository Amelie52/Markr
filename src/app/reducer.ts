import { Action } from "./actions";
import { Bookmark, Tag } from "./interfaces";

export type State = {
    bookmarks: Bookmark[];
    tags: Tag[];
}

export const initialState = {
    bookmarks: [],
    tags: [],
};

export function mainReducer(state: State = initialState, action: Action): State {
    switch (action.type) {
        case 'app/BOOKMARK/CREATE':
            return { ...state, bookmarks: [...state.bookmarks, { ...action.bookmark, tags: action.tags }] };
        case 'app/BOOKMARK/EDIT':
            return {
                ...state,
                bookmarks: state.bookmarks.map((bookmark) => {
                    if (bookmark.id === action.bookmarkId) {
                        return {
                            ...bookmark,
                            tags: action.tags
                        };
                    }
                    return bookmark;
                }),
            };
        case 'app/BOOKMARK/DELETE':
            return {
                ...state,
                bookmarks: state.bookmarks.filter((bookmark) => {
                    return bookmark.id !== action.bookmarkId;
                }),
            };
        case 'app/TAGS/CREATE':
            const newTags = action.tags.filter(tag => {
                return !state.tags.some(e => tag.value === e.value)
            })

            return {
                ...state,
                tags: [...state.tags, ...newTags]
            };
        default:
            throw new Error();
    }
}