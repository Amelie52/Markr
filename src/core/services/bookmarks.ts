import { Bookmark } from "../../app/interfaces";

export interface ReceivedBookmark {
    url: string;
    title: string;
    author_name?: string;
    width: number;
    height: number;
    duration?: number;
}

export function getFormattedBookmark(bookmark: ReceivedBookmark): Bookmark {
    // simulate an ID
    let uniqueId = Date.now().toString(36) + Math.random().toString(36).substring(2);
    const myDate = new Date();

    return {
        id: uniqueId,
        url: bookmark.url,
        title: bookmark.title,
        author: bookmark.author_name,
        width: bookmark.width,
        height: bookmark.height,
        duration: bookmark.duration,
        createdDate: myDate,
        tags: [],
    }
}

export function create({ url }: { url: string }) {
    return fetch(`https://noembed.com/embed?url=${url}`)
        .then(response => {
            return response.json();
        })
        .then(responseData => {
            return getFormattedBookmark(responseData)
        })
        .catch(error => {
            throw new Error('Il y a eu un problème avec l\'opération fetch: ' + error.message);
        });
}
