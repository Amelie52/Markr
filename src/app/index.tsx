import { createContext, Dispatch, useMemo, useReducer } from 'react';
import { BrowserRouter as Router, Route, Switch } from 'react-router-dom';
import styled, { createGlobalStyle, DefaultTheme, ThemeProvider } from 'styled-components';

import logo from '../images/logo.png';
import { State, mainReducer, initialState } from './reducer';
import Create from '../create';
import Edit from '../edit';
import Home from '../home';
import { Action } from './actions';

const GlobalStyle = createGlobalStyle`
  * {
    box-sizing: border-box;
  }

  html {
    font-size: .625em;
    font-size: calc(1em * .625);
  }
  
  body {
    margin: 0;
    padding: 0;
    font-size: ${props => props.theme.text.FONT_SIZE_MEDIUM};
    font-weight: 400;
    line-height: ${props => props.theme.text.LINE_HEIGHT};
  }

  #root {
    min-height: 100vh;
    display: flex;
    flex-direction: column;
  }

  a {
    color: black;
  }

  button, a {
    cursor: pointer;
    font-size: ${props => props.theme.text.FONT_SIZE_MEDIUM};
    line-height: ${props => props.theme.text.LINE_HEIGHT};
  }

  body, button {
    font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', 'Roboto', 'Oxygen',
    'Ubuntu', 'Cantarell', 'Fira Sans', 'Droid Sans', 'Helvetica Neue',
    sans-serif;
    -webkit-font-smoothing: antialiased;
    -moz-osx-font-smoothing: grayscale;
  }
`;

export const theme: DefaultTheme = {
  colors: {
    PRIMARY_COLOR: "#FD4586",
    SECONDARY_COLOR: "#fe7c65",
    GREY_MEDIUM: "#C4C4C4",
    GREY_DARK: "#AEAEAE",
    GREY_ULTRA_DARK: "#575757",
    GRADIENT: "linear-gradient(to left top, #FD4586, #FF9E8D);",
  },
  text: {
    FONT_SIZE_SMALL: "1.4rem", // 14px
    FONT_SIZE_MEDIUM: "1.6rem", // 16px
    FONT_SIZE_LARGE: "2rem", // 20px
    FONT_SIZE_XLARGE: "2.4rem", // 24px
    FONT_SIZE_XXLARGE: "3.2rem", // 32px
    LINE_HEIGHT: "1.5"
  },
  // breakpoints doesn't take html font-size
  breakpoints: {
    XX_SMALL: "26.25rem", // 420px
    X_SMALL: "36.25rem", // 580px
    SMALL: "46.875rem", // 750px
    MEDIUM: "56.25rem", // 900px
  },
  RADIUS: "0.25rem",
  BOX_SHADOW: "0px 0.125rem 0.8125rem 0.125rem rgba(196, 196, 196, 0.2)"
};

const Header = styled.header`
  padding: 1rem 2rem;
  z-index: 20;
  background-color: #fff;
  position: sticky;
  top: 0;
  box-shadow: ${props => props.theme.BOX_SHADOW};
`
const Main = styled.main`
  flex: 1 1 auto;
  display: flex;
  flex-direction: column;
`

const LogoContainer = styled.div`
  display: flex;
  align-items: center;
  font-size: 1.6rem;
  color: ${props => props.theme.colors.PRIMARY_COLOR};
`

const Logo = styled.img`
  width: 16px;
  height: 16px;
  margin-right: 1rem;
`

const Footer = styled.footer`
  padding: 1rem;
  color: ${props => props.theme.colors.GREY_ULTRA_DARK};
  text-align: center;
`

interface ContextProps {
  state: State;
  dispatch: Dispatch<Action>;
}

export const MainContext = createContext({} as ContextProps);

export const App = () => {
  const [state, dispatch] = useReducer(mainReducer, initialState);
  const date = new Date();

  const contextValue = useMemo(() => {
    return { state, dispatch };
  }, [state, dispatch]);

  return (
    <MainContext.Provider value={contextValue as ContextProps}>
      <ThemeProvider theme={theme}>
        <GlobalStyle />
        <Header role="banner">
          <LogoContainer><Logo src={logo} alt="Markr logo" />Markr</LogoContainer>
        </Header>
        <Main role="main">
          <Router>
            <Switch>
              <Route exact path="/">
                <Home />
              </Route>
              <Route path="/add">
                <Create />
              </Route>
              <Route path="/edit/:bookmarkId">
                <Edit />
              </Route>
            </Switch>
          </Router>
        </Main>
        <Footer>
          {`Â© ${date.getFullYear()} Markr`}
        </Footer>
      </ThemeProvider>
    </MainContext.Provider>
  );
}

export default App;
