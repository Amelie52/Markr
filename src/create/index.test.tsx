import { createMemoryHistory } from "history";
import { Router } from 'react-router-dom';

import { render, fireEvent, waitFor } from '../config/jest/testing-utils';
import Create from '.';
import * as bookmarksService from '../core/services/bookmarks';
import { MainContext } from '../app';
import { BOOKMARK, TAGS } from '../app/actions';
import { Tag } from '../app/interfaces';

jest.mock('../components/bookmarkForm', () => () => <div data-mock="bookmark-form-page" />);
jest.mock('../components/bookmarkForm', () => ({ onSubmit }: { onSubmit: (url: string, tags: Tag[]) => void }) => (
    <div data-mock="bookmark-form">
        <button data-testid="bookmark-form-submit-button" type="button" onClick={() => onSubmit('https://www.fake-test-url/video', [{ label: 'dolor', value: 'dolor' }])}>
            Submit
      </button>
    </div>
));

describe('src/create', () => {
    const dispatch = jest.fn();
    const state = {
        bookmarks: [],
        tags: [],
    };

    it('should renders create page', () => {
        const { container } = render(<Create />);
        expect(container).toMatchSnapshot();
    });

    it('should submit correctly', async () => {
        const spy = jest.spyOn(bookmarksService, 'create')
            .mockResolvedValueOnce({
                id: 'hihoi8799',
                url: 'https://www.fake-test-url/video',
                author: "amelie",
                title: 'The new vimeo player',
                width: 398,
                height: 879,
                duration: 76,
                tags: [{ label: 'lorem', value: 'lorem' }, { label: 'dolor', value: 'dolor' }],
                createdDate: new Date('2020-10-15T12:00:00.000Z')
            });

        const history = createMemoryHistory();
        history.push('/add')

        const { getByTestId } = render(
            <MainContext.Provider value={{ state, dispatch }}>
                <Router history={history}>
                    <Create />
                </Router>
            </MainContext.Provider>
        );

        const submitButton = getByTestId('bookmark-form-submit-button');
        fireEvent.click(submitButton);

        await waitFor(() => {
            expect(spy).toHaveBeenCalledWith({ url: 'https://www.fake-test-url/video' });

            expect(dispatch).toHaveBeenCalledTimes(2);
            expect(dispatch.mock.calls[0][0]).toEqual({
                type: BOOKMARK.CREATE,
                bookmark: {
                    id: 'hihoi8799',
                    url: 'https://www.fake-test-url/video',
                    author: "amelie",
                    title: 'The new vimeo player',
                    width: 398,
                    height: 879,
                    duration: 76,
                    tags: [{ label: 'lorem', value: 'lorem' }, { label: 'dolor', value: 'dolor' }],
                    createdDate: new Date('2020-10-15T12:00:00.000Z')
                },
                tags: [{ label: 'dolor', value: 'dolor' }],
            });
            expect(dispatch.mock.calls[1][0]).toEqual({
                type: TAGS.CREATE,
                tags: [{ label: 'dolor', value: 'dolor' }],
            });

            expect(history.location.pathname).toBe('/');
        })
    });

});
