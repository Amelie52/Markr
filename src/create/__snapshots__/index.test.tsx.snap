// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/create should renders create page 1`] = `
<div>
  <div
    class="sc-hKgILt PnehK"
  >
    <h1
      class="sc-eCssSg kLLfpA"
    >
      Ajouter un lien en favoris
    </h1>
    <button
      class="sc-dlfnbm bpTVga"
    >
      Revenir sur la page précédente
    </button>
    <div
      data-mock="bookmark-form"
    >
      <button
        data-testid="bookmark-form-submit-button"
        type="button"
      >
        Submit
      </button>
    </div>
  </div>
</div>
`;
